namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
 
 using namespace std;
 
+// TODO:
+// - Remove redundancy whenever possible.
+// - ENFORCE that NameFinder never produces errors.
+
 namespace sorbet::namer {
 
+namespace {
+class FoundNames;
+
+struct FoundClassRef;
+struct FoundClass;
+struct FoundStaticField;
+struct FoundTypeMember;
+struct FoundMethod;
+
+enum class NameKind : u1 {
+    None = 0,
+    Root = 1,
+    Class = 2,
+    ClassRef = 3,
+    Method = 4,
+    StaticField = 5,
+    TypeMember = 6,
+    Symbol = 7,
+};
+
+class FoundNameRef final {
+    u4 _id;
+
+    // Store NameKind in upper 8 bits of idx.
+    static u4 mask(NameKind kind, u4 idx) {
+        return (u4)kind << 28 | idx;
+    }
+
+public:
+    FoundNameRef() : FoundNameRef(NameKind::None, 0) {}
+    FoundNameRef(const FoundNameRef &nm) = default;
+    FoundNameRef(FoundNameRef &&nm) = default;
+    FoundNameRef &operator=(const FoundNameRef &rhs) = default;
+
+    FoundNameRef(NameKind kind, u4 idx) : _id(mask(kind, idx)) {}
+
+    static FoundNameRef root() {
+        return FoundNameRef(NameKind::Root, 0);
+    }
+
+    NameKind kind() const {
+        return (NameKind)(_id >> 28);
```

Comment:
Strong preference for `static_cast<NameKind>(...)` and `static_cast<u1>(...)` over using C-style casts:

https://stackoverflow.com/questions/28002/regular-cast-vs-static-cast-vs-dynamic-cast

Though it sounds like dmitry has asked whether the bit packing is worth it, in which case we wouldn't need casts at all.